### PR TITLE
Adding support for custom ValidationResult (derived class) in CustomV…

### PR DESCRIPTION
**Pull request by Liero**
#33767

### CODE REVIEW

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CustomValidationAttribute.cs [hunk 1]
```diff
@@ -242,8 +242,8 @@ private string ValidateMethodParameter()
                 return SR.Format(SR.CustomValidationAttribute_Method_Not_Found, Method, ValidatorType.Name);
             }
 
-            // Method must return a ValidationResult
-            if (methodInfo.ReturnType != typeof(ValidationResult))
+            // Method must return a ValidationResult or derived class
+             if (!typeof(ValidationResult).IsAssignableFrom(methodInfo.ReturnType))
```
Comment (Wraith2) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CustomValidationAttribute.cs]: Nit, remove the extra leading space

**Comment by Wraith2** [comment]
I don't think the CI failure is caused by your change. It's the usual failure to upload to helix on a single step but i don't see an actual test failure. relevant log items are:
```
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19126.3\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.ComponentModel.Annotations.Tests in job 1cf89d74-cf63-4328-b82f-8791e5f48a51 has Failed with exit code 1 [D:\a\1\s\eng\sendtohelix.proj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19126.3\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Job '1cf89d74-cf63-4328-b82f-8791e5f48a51' failed 1 out of 226855 tests. [D:\a\1\s\eng\sendtohelix.proj]
```
/cc @safern for confirmation that i'm reading that right.

If if is a false failure then you need to copy in the [area owners](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md) who are so they can review, @lajones, @divega, @ajcvickers

**Comment by safern** [comment]
This actually seems like a failure with this PR itself. If you look at the failure: 
`System.InvalidOperationException : The CustomValidationAttribute method 'ValidationMethodDerivedReturnTypeReturnsSomeError' in type 'CustomValidator' must return System.ComponentModel.DataAnnotations.ValidationResult.  Use System.ComponentModel.DataAnnotations.ValidationResult.Success to represent success.`

It is failing directly here, on the added CustomValidator. It makes sense, since NETFX doesn't have this support: https://github.com/dotnet/corefx/pull/35584/files#diff-5fb1c2fa77569a06289a2ffca62c378aR246

So the test will need to be conditioned to only run on netcoreapp.

To skip the test entirely for NETFX, we can add an attribute to https://github.com/dotnet/corefx/pull/35584/files#diff-80b5d2f008710e7c2a1b943f176c98dbR183:
`[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Reason to skip it")`]

**Comment by Wraith2** [comment]
Ok, I can't work out where you're seeing that, perhaps i'm missing something obvious. If I click though the netfx coreci failure Details link above, then click the errors/warnings link and I get to here:
capture
which only allows me to see the helix upload step is failing and the two messages above, I can't find the errors you're seeing anywhere.

**Comment by safern** [comment]
If you click on the `Send to Helix` step to open the logs, then there will be a URL within the log, something like:
`Results will be available from https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F35584~2Fmerge/test~2Ffunctional~2Fcli~2F/20190227.1`

image

**Comment by Wraith2** [comment]
Oh! ok, so you follow the links through all the fancy stuff to get a link that you have to copy paste out into a totally different app. That's not quite as obvious as it could perhaps be. Thanks for the info though, that'll stop me copying you into issues so you can read the log for me 😀

**Comment by safern** [comment]
> Oh! ok, so you follow the links through all the fancy stuff to get a link that you have to copy paste out into a totally different app. That's not quite as obvious as it could perhaps be. Thanks for the info though, that'll stop me copying you into issues so you can read the log for me 😀

You can also do CTRL+Click and you don't have to copy the link from the logs. We're definitely working on improving that experience since it is painful, but we're limited now since we depend from azure devops fixes we requested, but it should improve on time :smile:

**Comment by Liero** [comment]
I'm not sure I follow why it is failing in .NET Framework..

**Comment by safern** [comment]
> I'm not sure I follow why it is failing in .NET Framework..

Because .NET Framework doesn't support custom ValidationResult, because the support you're adding here is for .NET Core only, bits for .NET Core and UAP from this repo. .NET Framework is another code base and most likely it will not get this back-ported.

**Comment by Liero** [comment]
Ok, I will fix the tests. However, is there some article that explains how this works, how to test in on local machine etc? Some guidance for community. Or I'm left only with the readmes in this repo?

**Comment by Wraith2** [comment]
Most of what you need it in the documentation folder in various places. There's also a doc just being added which covers some of the newer aspects https://github.com/dotnet/corefx/pull/35672#discussion_r261471673

**Comment by ajcvickers** [comment]
Thanks!

**Pull request merged**